### PR TITLE
Add support source-ip-address to resource_sql_database_instance

### PR DESCRIPTION
**Pull request by paulpalamarchuk**
* Fixes https://github.com/terraform-providers/terraform-provider-google/issues/3704
```
`sql`: added `source_ip_address` and `source_port` fields to `google_sql_database_instance` resource

### CODE REVIEW

**Comment by danawillow** [comment]
Hey @paulpalamarchuk, thanks for the PR! I noticed you opened it as a draft- are you looking for any feedback now, or should I hold off until this is ready?

**Comment by paulpalamarchuk** [comment]
@danawillow I'll complete some tests and update PR if needed. Then will move it to "ready for review" state.

**Comment by danawillow** [comment]
Sounds good! I'm not sure if GitHub emails on changes from Draft -> Ready, so if you wouldn't mind also commenting that'll help make sure I don't miss it.

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 1]
```diff
@@ -78,6 +78,20 @@ var (
 		"replica_configuration.0.username",
 		"replica_configuration.0.verify_server_certificate",
 	}
+<% unless version == 'ga' -%>
+	externalReplicaConflictKeys = []string{
+  	"settings",
```
Comment (kopachevsky) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: strange tab here

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 2]
```diff
@@ -333,11 +351,27 @@ func resourceSqlDatabaseInstance() *schema.Resource {
...
+				Optional:      true,
+				ConflictsWith: externalReplicaConflictKeys,
+				ValidateFunc:  validation.SingleIP(),
+			},
+			"source_port": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: externalReplicaConflictKeys,
+				Default:       "3306",
+				},
```
Comment (kopachevsky) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: formatting

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 3]
```diff
@@ -333,11 +351,27 @@ func resourceSqlDatabaseInstance() *schema.Resource {
...
 			<% end -%>
 
 			"ip_address": {
 				Type:     schema.TypeList,
+				<% if version == 'ga' -%>
 				Computed: true,
+				<% else %>
+				Optional: true,
+				<% end -%>
 				Elem: &schema.Resource{
```
Comment (kopachevsky) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: space before {

**Comment by paulpalamarchuk** [comment]
@danawillow PR is ready for review. Please take a look.

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 4]
```diff
@@ -333,12 +351,27 @@ func resourceSqlDatabaseInstance() *schema.Resource {
...
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: externalReplicaConflictKeys,
+				ValidateFunc:  validation.SingleIP(),
+			},
+			"source_port": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: externalReplicaConflictKeys,
+				InputDefault:  "3306",
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: We don't use `InputDefault` anywhere else in the provider- is there a reason this shouldn't just be `Default`?

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 5]
```diff
@@ -333,12 +351,27 @@ func resourceSqlDatabaseInstance() *schema.Resource {
...
+			},
 			<% end -%>
 
 			"ip_address": {
 				Type:     schema.TypeList,
 				Computed: true,
-				Elem: &schema.Resource{
+				<% unless version == 'ga' -%>
+                                Optional: true,
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: nit: should be tabs

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 6]
```diff
@@ -403,42 +436,63 @@ func resourceSqlDatabaseInstance() *schema.Resource {
 						"ca_certificate": {
 							Type:         schema.TypeString,
 							Optional:     true,
+							<% unless version == 'ga' -%>
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: I'm curious why these are different between beta and GA

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 7]
```diff
@@ -561,25 +632,46 @@ func resourceSqlDatabaseInstanceCreate(d *schema.ResourceData, meta interface{})
...
 	<% unless version == 'ga' -%>
 	// MSSQL Server require rootPassword to be set
 	if strings.Contains(instance.DatabaseVersion, "SQLSERVER") {
 		instance.RootPassword = d.Get("root_password").(string)
 	}
+
+	if sqlDatabaseIsOnPremises(d) {
+		instance.OnPremisesConfiguration = &sqladmin.OnPremisesConfiguration{
+			HostPort:        d.Get("source_ip_address").(string) + ":" + d.Get("source_port").(string),
+			Kind:            "sql#onPremisesConfiguration",
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: `Kind` usually doesn't actually need to be set

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 8]
```diff
@@ -561,25 +632,46 @@ func resourceSqlDatabaseInstanceCreate(d *schema.ResourceData, meta interface{})
...
 
 	<% unless version == 'ga' -%>
 	// MSSQL Server require rootPassword to be set
 	if strings.Contains(instance.DatabaseVersion, "SQLSERVER") {
 		instance.RootPassword = d.Get("root_password").(string)
 	}
+
+	if sqlDatabaseIsOnPremises(d) {
+		instance.OnPremisesConfiguration = &sqladmin.OnPremisesConfiguration{
+			HostPort:        d.Get("source_ip_address").(string) + ":" + d.Get("source_port").(string),
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: Instead of having two fields that we then combine, what if we just matched the API here so you had an `on_premises_configuration` block with a `host_port` property?

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 9]
```diff
@@ -812,46 +908,63 @@ func resourceSqlDatabaseInstanceRead(d *schema.ResourceData, meta interface{}) e
...
-	d.Set("connection_name", instance.ConnectionName)
-	d.Set("service_account_email_address", instance.ServiceAccountEmailAddress)
+	<% unless version == 'ga' -%>
+	// Set external master config if there is any.
+	if instance.OnPremisesConfiguration != nil {
+		d.Set("source_ip_address", (strings.Split(instance.OnPremisesConfiguration.HostPort, ":"))[0])
+		d.Set("source_port", (strings.Split(instance.OnPremisesConfiguration.HostPort, ":"))[1])
+	} else {
+	<% end -%>
+	        d.Set("connection_name", instance.ConnectionName)
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: nit: should be tabs

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 10]
```diff
@@ -812,46 +908,63 @@ func resourceSqlDatabaseInstanceRead(d *schema.ResourceData, meta interface{}) e
...
+		d.Set("service_account_email_address", instance.ServiceAccountEmailAddress)
 
-	if err := d.Set("replica_configuration", flattenReplicaConfiguration(instance.ReplicaConfiguration, d)); err != nil {
-		log.Printf("[WARN] Failed to set SQL Database Instance Replica Configuration")
-	}
-	ipAddresses := flattenIpAddresses(instance.IpAddresses)
-	if err := d.Set("ip_address", ipAddresses); err != nil {
-		log.Printf("[WARN] Failed to set SQL Database Instance IP Addresses")
-	}
+		if err := d.Set("settings", flattenSettings(instance.Settings)); err != nil {
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: Looks like this block is here twice

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 11]
```diff
@@ -812,46 +908,63 @@ func resourceSqlDatabaseInstanceRead(d *schema.ResourceData, meta interface{}) e
...
 	d.Set("region", instance.Region)
 	d.Set("database_version", instance.DatabaseVersion)
-	d.Set("connection_name", instance.ConnectionName)
-	d.Set("service_account_email_address", instance.ServiceAccountEmailAddress)
+	<% unless version == 'ga' -%>
+	// Set external master config if there is any.
+	if instance.OnPremisesConfiguration != nil {
+		d.Set("source_ip_address", (strings.Split(instance.OnPremisesConfiguration.HostPort, ":"))[0])
+		d.Set("source_port", (strings.Split(instance.OnPremisesConfiguration.HostPort, ":"))[1])
+	} else {
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: It's generally a good idea to set things even if they're empty, so I would probably have this `else` only contain `d.Set`s for the on prem properties and then have the other stuff get set unconditionally

File: third_party/terraform/website/docs/r/sql_database_instance.html.markdown [hunk 12]
```diff
@@ -217,6 +217,10 @@ instances support `MYSQL_5_5` or `MYSQL_5_6`.
     
 * `root_password` - (Optional, [Beta](https://terraform.io/docs/providers/google/guides/provider_versions.html)) Initial root password. Required for MS SQL Server, ignored by MySQL and PostgreSQL.
 
+* `source_ip_address` - (Optional, [Beta](https://www.terraform.io/docs/providers/google/guides/provider_versions.html#google-beta)) Public IP address used to connect to and replicate from the [external `MySQL` data source](https://cloud.google.com/sql/docs/mysql/replicatiohttps://cloud.google.com/sql/docs/mysql/replication/replication-from-external#setupn/replication-from-external#setup). Conflicts with `settings` block supports.
```
Comment (danawillow) [inline third_party/terraform/website/docs/r/sql_database_instance.html.markdown]:
```suggestion
* `source_ip_address` - (Optional, [Beta](https://www.terraform.io/docs/providers/google/guides/provider_versions.html#google-beta)) Public IP address used to connect to and replicate from the [external `MySQL` data source](https://cloud.google.com/sql/docs/mysql/replication/replication-from-external#setupn/replication-from-external#setup). Cannot be set if the `settings` block is set.
```

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 13]
```diff
@@ -78,6 +78,20 @@ var (
 		"replica_configuration.0.username",
 		"replica_configuration.0.verify_server_certificate",
 	}
+<% unless version == 'ga' -%>
+	externalReplicaConflictKeys = []string{
+		"settings",
+		"replica_configuration",
+		"ip_address",
+		"server_ca_cert",
+		"connection_name",
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: since `connection_name` is computed, it can't be set by users anyway, so it doesn't really need to be here (likewise for a few of the others)

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 14]
```diff
@@ -473,31 +536,39 @@ func resourceSqlDatabaseInstance() *schema.Resource {
 			"server_ca_cert": {
 				Type:     schema.TypeList,
 				Computed: true,
+				<% unless version == 'ga' -%>
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: This change doesn't seem relevant to this PR- is it?

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 15]
```diff
@@ -561,25 +632,46 @@ func resourceSqlDatabaseInstanceCreate(d *schema.ResourceData, meta interface{})
...
 	if strings.Contains(instance.DatabaseVersion, "SQLSERVER") {
 		instance.RootPassword = d.Get("root_password").(string)
 	}
+
+	if sqlDatabaseIsOnPremises(d) {
+		instance.OnPremisesConfiguration = &sqladmin.OnPremisesConfiguration{
+			HostPort:        d.Get("source_ip_address").(string) + ":" + d.Get("source_port").(string),
+			Kind:            "sql#onPremisesConfiguration",
+		}
+	} else {
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: In general, I'd prefer to use as little business logic as possible in here. If we send empty values in the request they'll get ignored, so it's easier to just send them unconditionally and let the client library take care of removing the fields that aren't set.

File: third_party/terraform/website/docs/r/sql_database_instance.html.markdown (hunk 12 above)
Comment (danawillow) [inline third_party/terraform/website/docs/r/sql_database_instance.html.markdown]: Can you also update docs for any other fields that you changed as part of this PR?

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 4 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: `Default` property will always set that parameter, but we don`t need it, because of conflicts with `settings`

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 5 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: fixed

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 7 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: Removed

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 8 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: The idea is to set only IP address, if the port is standard, in modules, that will use that resource.

File: third_party/terraform/website/docs/r/sql_database_instance.html.markdown (hunk 12 above)
Comment (paulpalamarchuk) [reply third_party/terraform/website/docs/r/sql_database_instance.html.markdown]: Updated

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 13 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: removed

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 14 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: I've tried to avoid diff in `state`

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 8 above)
Comment (danawillow) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: I'd rather not add logic at the provider level that's specific to a certain module use-case. That feels better suited for the module itself.

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 14 above)
Comment (danawillow) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: Why was there a diff in state for a computed field?

**Comment by paulpalamarchuk** [comment]
@danawillow I've updated the PR. Please take a look. Notify me, if any additional changes nedded.

**Comment by danawillow** [comment]
It looks like there are a few unresolved comments still from my last review- can you double-check those?

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 6 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: Removed

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 8 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: Reworked

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 14 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: I've re-checked, it's not needed. Removed

File: third_party/terraform/website/docs/r/sql_database_instance.html.markdown (hunk 12 above)
Comment (paulpalamarchuk) [reply third_party/terraform/website/docs/r/sql_database_instance.html.markdown]: Changed description in other fields

**Comment by danawillow** [comment]
Just wanted to poke my head back in here to let you know that I haven't forgotten about this, but my week has been on the busier side so I may not get to it until Monday

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 15 above)
Comment (danawillow) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: ^

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 11 above)
Comment (danawillow) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: ^

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 16]
```diff
@@ -812,45 +856,54 @@ func resourceSqlDatabaseInstanceRead(d *schema.ResourceData, meta interface{}) e
...
+		d.Set("private_ip_address", privateIpAddress)
+		d.Set("master_instance_name", strings.TrimPrefix(instance.MasterInstanceName, project+":"))
+<% unless version == 'ga' -%>
+	}
+<% end -%>
 
 	if err := d.Set("server_ca_cert", flattenServerCaCert(instance.ServerCaCert)); err != nil {
-		log.Printf("[WARN] Failed to set SQL Database CA Certificate")
-	}
+        	log.Printf("[WARN] Failed to set SQL Database CA Certificate")
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: can you change this back to tabs?

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 17]
```diff
@@ -866,10 +919,18 @@ func resourceSqlDatabaseInstanceUpdate(d *schema.ResourceData, meta interface{})
 		return err
 	}
 
-	// Update only updates the settings, so they are all we need to set.
-	instance := &sqladmin.DatabaseInstance{
-		Settings: expandSqlDatabaseInstanceSettings(d.Get("settings").([]interface{}), !isFirstGen(d)),
+	instance := &sqladmin.DatabaseInstance{}
+	<% unless version == 'ga' -%>
+	if sqlDatabaseIsOnPremises(d) {
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: same comment re: if/elses. I'd rather just set things and allow the client library to filter out nils. You can add an expander for OnPremisesConfiguration that'll return nil if it isn't set, similarly to how we do for settings.

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 18]
```diff
@@ -333,12 +345,20 @@ func resourceSqlDatabaseInstance() *schema.Resource {
 				ForceNew:  true,
 				Sensitive: true,
 			},
+			"source_host_port": {
```
Comment (danawillow) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb]: If GCP adds an additional field to the OnPremisesConfiguration block, then we're in an awkward situation in the provider by having broken this one out into its own field. Making this match the API by putting it in a block (in TF schema it would be a TypeList with MaxItems 1) is more future-proof.

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 11 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: I've tested - it doesn't work that way. Getting error:
```
Error: Error, failed to create instance tf-test-external-master-rmk1pcfeil: googleapi: Error 400: Missing parameter: Tier., required
```

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 18 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: Changed it to the block

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 15 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: I've tested - it doesn't work that way. Getting error:
```
Error: Error, failed to create instance tf-test-external-master-rmk1pcfeil: googleapi: Error 400: Missing parameter: Tier., required
```

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 16 above)
Comment (paulpalamarchuk) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb]: Changed

**Comment by danawillow** [comment]
In reviewing this, I noticed that a different open PR had written this as a separate resource, which I think is a really good idea. I'm going to go ahead and close this PR out, but prioritize getting this feature in soon. If it's important for you to be the author let me know and I'll either reopen this or you can propose a new PR, otherwise someone from our team will work on it within the next ~2 weeks. There's some additional context at https://github.com/terraform-providers/terraform-provider-google/issues/5620#issuecomment-589799193, and thanks for your patience and understanding!
